File: src/test/java/no/nav/familie/integrasjoner/tilgangskontroll/TilgangskontrollServiceTest.java
package no.nav.familie.integrasjoner.tilgangskontroll;

import no.nav.familie.integrasjoner.azure.domene.Saksbehandler;
import no.nav.familie.integrasjoner.egenansatt.EgenAnsattService;
import no.nav.familie.integrasjoner.personopplysning.domene.PersonIdent;
import no.nav.familie.integrasjoner.personopplysning.domene.Personinfo;
import no.nav.familie.integrasjoner.tilgangskontroll.domene.Tilgang;
import no.nav.familie.integrasjoner.azure.AzureGraphService;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.junit4.SpringRunner;

import java.time.LocalDate;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@RunWith(SpringRunner.class)
public class TilgangskontrollServiceTest {

    private static final String NAVN = "Navn Navnesen";
    private static final String FNR = "fnr";
    private static final PersonIdent PERSON_IDENT = new PersonIdent(FNR);

    @MockBean
    private EgenAnsattService egenAnsattService;
    private TilgangsKontrollService tilgangsKontrollService;

    @MockBean
    private AzureGraphService azureGraphService;
    private Saksbehandler saksbehandler = new Saksbehandler("", "sdfsdf");


    @Before
    public void setUp() {
        tilgangsKontrollService = new TilgangsKontrollService(azureGraphService, egenAnsattService);
    }

    @Test
    public void tilgang_til_egen_ansatt_gir_ikke_tilgang_hvis_saksbehandler_mangler_rollen() {
        when(egenAnsattService.erEgenAnsatt(any(String.class))).thenReturn(Boolean.TRUE);

        assertThat(tilgangsKontrollService.sjekkTilgang("123", saksbehandler, this.personinfoUtenKode6og7()).isHarTilgang())
                .isEqualTo(new Tilgang().withHarTilgang(false).isHarTilgang());
    }

    @Test
    public void tilgang_til_egen_ansatt_gir_OK_hvis_søker_ikke_er_egen_ansatt() {
        when(egenAnsattService.erEgenAnsatt(any(String.class))).thenReturn(Boolean.FALSE);

        assertThat(tilgangsKontrollService.sjekkTilgang("123", saksbehandler, this.personinfoUtenKode6og7()).isHarTilgang())
                .isEqualTo(new Tilgang().withHarTilgang(true).isHarTilgang());
    }

    @Test
    public void hvis_kode6_har_ikke_saksbehandler_tilgang() {
        when(egenAnsattService.erEgenAnsatt(any(String.class))).thenReturn(Boolean.FALSE);

        assertThat(tilgangsKontrollService.sjekkTilgang("123", saksbehandler, this.personinfoMedKode6()).isHarTilgang())
                .isEqualTo(new Tilgang().withHarTilgang(false).isHarTilgang());
    }

    @Test
    public void hvis_kode7_har_ikke_saksbehandler_tilgang() {
        when(egenAnsattService.erEgenAnsatt(any(String.class))).thenReturn(Boolean.FALSE);

        assertThat(tilgangsKontrollService.sjekkTilgang("123", saksbehandler, this.personinfoMedKode7()).isHarTilgang())
                .isEqualTo(new Tilgang().withHarTilgang(false).isHarTilgang());
    }

    private Personinfo personinfoUtenKode6og7() {
        return new Personinfo.Builder().medPersonIdent(PERSON_IDENT)
                                       .medFødselsdato(LocalDate.now())
                                       .medNavn(NAVN)
                                       .medDiskresjonsKode("VANLIG_PERSON")
                                       .build();
    }

    private Personinfo personinfoMedKode6() {
        return new Personinfo.Builder().medPersonIdent(PERSON_IDENT)
                                       .medFødselsdato(LocalDate.now())
                                       .medNavn(NAVN)
                                       .medDiskresjonsKode(TilgangsKontrollService.DISKRESJONSKODE_KODE6)
                                       .build();
    }
    private Personinfo personinfoMedKode7() {
        return new Personinfo.Builder().medPersonIdent(PERSON_IDENT)
                                       .medFødselsdato(LocalDate.now())
                                       .medNavn(NAVN)
                                       .medDiskresjonsKode(TilgangsKontrollService.DISKRESJONSKODE_KODE7)
                                       .build();
    }
}

File: src/test/java/no/nav/familie/integrasjoner/tilgangskontroll/TilgangskontrollServiceTest.kt
package no.nav.familie.integrasjoner.tilgangskontroll

import io.mockk.every
import io.mockk.mockk
import no.nav.familie.integrasjoner.config.TilgangConfig
import no.nav.familie.integrasjoner.egenansatt.EgenAnsattService
import no.nav.familie.integrasjoner.personopplysning.PersonopplysningerService
import no.nav.familie.integrasjoner.personopplysning.internal.ADRESSEBESKYTTELSEGRADERING
import no.nav.familie.integrasjoner.personopplysning.internal.ADRESSEBESKYTTELSEGRADERING.FORTROLIG
import no.nav.familie.integrasjoner.personopplysning.internal.ADRESSEBESKYTTELSEGRADERING.STRENGT_FORTROLIG
import no.nav.familie.integrasjoner.personopplysning.internal.ADRESSEBESKYTTELSEGRADERING.UGRADERT
import no.nav.familie.integrasjoner.personopplysning.internal.Adressebeskyttelse
import no.nav.familie.integrasjoner.personopplysning.internal.Person
import no.nav.familie.integrasjoner.tilgangskontroll.domene.AdRolle
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken

class TilgangskontrollServiceTest {
    private val egenAnsattService: EgenAnsattService = mockk(relaxed = true)

    private val tilgangConfig: TilgangConfig =
        TilgangConfig(
            egenAnsatt = AdRolle(GRUPPE_EGEN_ANSATT, "NAV-Ansatt"),
            kode6 = AdRolle(GRUPPE_TILGANG_6, "Strengt fortrolig adresse"),
            kode7 = AdRolle(GRUPPE_TILGANG_7, "Fortrolig adresse"),
        )
    private val personopplysningerService: PersonopplysningerService = mockk(relaxed = true)

    private val cachedTilgangskontrollService =
        CachedTilgangskontrollService(
            egenAnsattService,
            personopplysningerService,
            tilgangConfig,
        )
    private var tilgangskontrollService: TilgangskontrollService = TilgangskontrollService(cachedTilgangskontrollService)

    @BeforeEach
    fun setup() {
        mockToken(tilganger = emptyList())
    }

    @AfterEach
    fun cleanUp() {
        SecurityContextHolder.clearContext()
    }

    @Test
    fun `tilgang til egen ansatt gir ikke tilgang hvis saksbehandler mangler rollen`() {
        every { egenAnsattService.erEgenAnsatt(any<String>()) } returns true

        assertThat(tilgangskontrollService.sjekkTilgang("123").harTilgang).isFalse
    }

    @Test
    fun `tilgang til egen ansatt gir tilgang hvis saksbehandler har rollen`() {
        every { egenAnsattService.erEgenAnsatt(any<String>()) } returns true
        every { personopplysningerService.hentAdressebeskyttelse("123", any()) } returns Adressebeskyttelse(UGRADERT)

        mockToken(tilganger = listOf(GRUPPE_EGEN_ANSATT))

        assertThat(tilgangskontrollService.sjekkTilgang("123").harTilgang).isTrue
    }

    @Test
    fun `tilgang til egen ansatt gir ok hvis søker ikke er egen ansatt`() {
        every { egenAnsattService.erEgenAnsatt(any<String>()) } returns false
        every { personopplysningerService.hentAdressebeskyttelse("123", any()) } returns Adressebeskyttelse(UGRADERT)

        assertThat(tilgangskontrollService.sjekkTilgang("123").harTilgang).isTrue
    }

    @Test
    fun `hvis kode6 har ikke saksbehandler uten rollen tilgang`() {
        every { egenAnsattService.erEgenAnsatt(any<String>()) } returns false
        every { personopplysningerService.hentPersoninfo("123", any()) } returns personMedAdressebeskyttelse(STRENGT_FORTROLIG)

        assertThat(tilgangskontrollService.sjekkTilgang("123").harTilgang).isFalse
    }

    @Test
    fun `hvis kode7 har ikke saksbehandler uten rollen tilgang`() {
        every { egenAnsattService.erEgenAnsatt(any<String>()) } returns false
        every { personopplysningerService.hentPersoninfo("123", any()) } returns personMedAdressebeskyttelse(FORTROLIG)

        assertThat(tilgangskontrollService.sjekkTilgang("123").harTilgang).isFalse
    }

    @Test
    fun `hvis kode6 har saksbehandler med rollen tilgang`() {
        every { egenAnsattService.erEgenAnsatt(any<String>()) } returns false
        every { personopplysningerService.hentPersoninfo("123", any()) } returns personMedAdressebeskyttelse(STRENGT_FORTROLIG)

        mockToken(tilganger = listOf(GRUPPE_TILGANG_6))

        assertThat(tilgangskontrollService.sjekkTilgang("123").harTilgang).isTrue
    }

    @Test
    fun `hvis kode7 har saksbehandler med rollen tilgang`() {
        every { egenAnsattService.erEgenAnsatt(any<String>()) } returns false
        every { personopplysningerService.hentAdressebeskyttelse("123", any()) } returns Adressebeskyttelse(FORTROLIG)

        mockToken(tilganger = listOf(GRUPPE_TILGANG_7))

        assertThat(tilgangskontrollService.sjekkTilgang("123").harTilgang).isTrue
    }

    private fun mockToken(tilganger: List<String>) {
        val jwt =
            Jwt
                .withTokenValue("token")
                .header("header", "header value")
                .claim("groups", tilganger)
                .build()
        val securityContext = SecurityContextHolder.createEmptyContext()
        securityContext.authentication = JwtAuthenticationToken(jwt)
        SecurityContextHolder.setContext(securityContext)
    }

    private fun personMedAdressebeskyttelse(adressebeskyttelsegradering: ADRESSEBESKYTTELSEGRADERING?): Person =
        Person(
            navn = "Navn Navnesen",
            adressebeskyttelseGradering = adressebeskyttelsegradering,
        )

    companion object {
        private const val GRUPPE_EGEN_ANSATT = "utvidetTilgang1"
        private const val GRUPPE_TILGANG_6 = "kode62"
        private const val GRUPPE_TILGANG_7 = "kode73"
    }
}
